apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data.IntegrationTests/V1/SessionPoolTests.cs

```diff
@@ -0,0 +1,111 @@
+                // Upper bound for when we expect we need to refresh.
+                var expectedRefreshTime = DateTime.UtcNow + options.IdleSessionRefreshDelay;
+                // Acquire all the sessions, which should all have a required refresh time within our upper bound
+                var initialSessions = await AcquireSessionsAsync(pool);
+                Assert.All(initialSessions, s => Assert.True(s.RefreshTimeForTest < expectedRefreshTime));
+
+                // Wait for everything to idle out
+                await Task.Delay(options.IdleSessionRefreshDelay + TimeSpan.FromSeconds(1));
+                Assert.True(DateTime.UtcNow > expectedRefreshTime);
+
+                // Release the sessions, wait for the pool to repopulate, then fetch again
+                initialSessions.ForEach(s => s.ReleaseToPool(false));
+                cts = new CancellationTokenSource(10000);
+                await pool.WaitForPoolAsync(_fixture.DatabaseName, cts.Token);
+                Logger.DefaultLogger.Info(pool.GetStatisticsSnapshot(_fixture.DatabaseName).ToString());
+                var refreshedSessions = await AcquireSessionsAsync(pool);
+
+                // We should still have the same sessions...
+                Assert.Equal(
+                    initialSessions.Select(s => s.SessionName.SessionId).OrderBy(id => id),
+                    refreshedSessions.Select(s => s.SessionName.SessionId).OrderBy(id => id));
+
+                // ... but they should all be refreshed now.
+                Assert.All(refreshedSessions, s => Assert.True(s.RefreshTimeForTest > expectedRefreshTime));
+                // Release the sessions so they can be deleted when we shut down the pool.
+                refreshedSessions.ForEach(s => s.ReleaseToPool(false));
+            });
+
+            async Task<List<PooledSession>> AcquireSessionsAsync(SessionPool pool)
+            {
+                var ret = new List<PooledSession>();
+                for (int i = 0; i < options.MinimumPooledSessions; i++)
+                {
+                    var cts = new CancellationTokenSource(1000);
+                    ret.Add(await pool.AcquireSessionAsync(_fixture.DatabaseName, null, cts.Token));
+                }
+                return ret;
+            }
+        }
+
+        private async Task TestWithPool(SessionPoolOptions options, Func<SessionPool, Task> test)
+        {
+            var poolManager = SessionPoolManager.Create(options);
+            var pool = await poolManager.AcquireSessionPoolAsync(_fixture.SpannerClientCreationOptions);
+            try
+            {
+                await test(pool);
+            }
+            finally
+            {
+                // Wait for up to 10 seconds for the pool to shut down, but don't fail if it doesn't. This would
+                // usually be because the test failed and didn't get it into a "clean shutdown" situation.
+                await Task.WhenAny(
```

Comment:
There's a helper method that does this (waits for 10 seconds for shutdown). I think you can use it here instead? It's in `SessionPoolHelpers`.

Reply:
That doesn't quite work, because we want to use the custom SessionPoolManager rather than the default one.
Also, that fails with a cancelled exception if it times out, whereas we want to just ignore it.

Reply:
Oh true.